desktop_version/src/FileSystemUtils.cpp

```diff
@@ -1042,26 +1042,120 @@ void FILESYSTEM_enumerateLevelDirFileNames(
     }
 }
 
-std::vector<std::string> FILESYSTEM_getLanguageCodes(void)
+typedef struct _enum_handle
 {
-    std::vector<std::string> list;
-    char** fileList = PHYSFS_enumerateFiles("lang");
+    char** physfs_list;
     char** item;
+    char* mounted_path;
+}
+enum_handle;
 
-    for (item = fileList; *item != NULL; item++)
+void FILESYSTEM_freeEnumerate(void** p_handle)
+{
+    /* Call this function after enumerating with FILESYSTEM_enumerate or friends. */
+    if (p_handle != NULL)
     {
-        char fullName[128];
-        SDL_snprintf(fullName, sizeof(fullName), "lang/%s", *item);
+        enum_handle* handle = (enum_handle*) *p_handle;
+        if (handle != NULL)
+        {
+            PHYSFS_freeList(handle->physfs_list);
+            VVV_free(handle->mounted_path);
+            VVV_free(handle);
+        }
+    }
+}
 
-        if (FILESYSTEM_isDirectory(fullName) && *item[0] != '.')
+const char* FILESYSTEM_enumerate(const char* folder, void** p_handle)
```

Comment:
Originally I did have the handle type as a public struct, but I wasn't really satisfied with the code until I made it private to `FileSystemUtils.cpp` like that - it resulted in many simplifications, off the top of my head:
- The name could be simply `enum_handle` instead of something like `FILESYSTEM_enum_handle`
- `MAX_PATH` was a big problem when I stored the mounted path in the struct, and one of the reasons for the change - originally I had a malloc for the path just because only `FileSystemUtils.cpp` knew about `MAX_PATH`, until I could change it into a fixed-size array by declaring it only in `FileSystemUtils.cpp`.
- I aimed to make the callsite code as simple as possible, to make it easier to use and to have less code duplication like always having to declare the same couple of variables. This is at the cost of a little more complexity in the implementation and a heap allocation, but it should be well-enough documented. I don't really like the idea of copying that `SDL_zero(handle);` to every callsite that just wants to enumerate files.
- If the callsite accesses `enum_handle.item` directly, it gives a very misleading result since it doesn't point to the current item. Though I guess some good naming could've helped, it felt like an implementation detail better not exposed.

Reply:
- The name can just be `EnumHandle`; it doesn't have to be something like `FILESYSTEM_enum_handle`.
- It can just be initialized like `enum_handle handle = {};`, then.
- Name it `_item` then. The leading underscore means it's an internal implementation detail that shouldn't be used by the caller (even though leading underscores are randomly used elsewhere in the codebase when they shouldn't be...).

Reply:
Hm, sounds like an acceptable compromise.

Reply:
Did all of this, and also made the enumeration stuff a single commit.